cassdegrees/api/degreeRules.py

```diff
@@ -1,52 +1,161 @@
             elif token == ")":
                 if len(stack) > 0:
                     node = stack.pop()
                 else:
-                    print("h")
-                    parent = {}
-                    parent["left"] = node
+                    parent = [None]
+                    parent.append(node)
                     node = parent
                     tree = parent
             elif token in ["AND", "OR"]:
-                node['val'] = token
-                node['right'] = {}
+                node[0] = token
+                node.append([None])
                 stack.append(node)
-                node = node['right']
+                node = node[-1]
             else:
-                node['rule'] = token
-                print(stack)
+                node[0] = token
                 parent = stack.pop()
-                print(stack)
                 node = parent
+
+        self.expressionTree = tree
+
         return tree
-        #create rules, return rules. Then take a list of rules in a bool alg expression and build the json
 
-    # Add rule as another AND?
+    def append_rule(self, rule):
+        """
+        Takes a rule that should be independent of the existing rules for a degree
+        and appends it to the degree rules with an AND operator.
+
+        :param rule: (Dict) the rule to be appended
+        """
+
+        if self.expressionTree[0] == "AND":
+            self.expressionTree.append(rule)
+        else:
+            self.expressionTree = ["AND", self.expressionTree, rule]
+
+    def get_rules_as_json(self):
+        """
+        :return: a JSON serialised string representing the rules expression tree
+        """
+
+        return json.dumps(self.expressionTree)
+
+    """
+        DEGREE RULES
+            Functions that return dictionary representations of rules
+    """
+
+    @staticmethod
+    def rule_course_selection(code):
+        """
+        Creates a rule which lists a single course code which
+        must be completed for a degree.
+
+        :param code: (String) the course code
+        :return: (Dict) a course selection rule
+        """
+
+        rule = {"type": "course selection", "code": code}
+        return rule
+
+    @staticmethod
+    def rule_subplan_selection(code):
+        """
+        Creates a rule which lists a single subplan code which
+        must be completed for a degree.
+
+        :param code: (String) the subplan code
+        :return: (Dict) a subplan selection rule
+        """
+
+        rule = {"type": "subplan selection", "code": code}
+        return rule
+
+    @staticmethod
+    def rule_electives(units):
+        """
+        Creates a rule which describes a number of units worth
+        of elective courses which must be completed for the degree.
+
+        :param units: (Int) number of units of available electives
+        :return: (Dict) an elective rule
+        """
+
+        rule = {"type": "electives", "units": units}
```

Comment:
This is fine for free electives, but there are no rules that allow you to generate list electives (24 units from courses A, B, C, D, E). We will need to have this in order to build more complex degrees

Reply:
Yep, it's an incomplete set of rules. This is just the base functionality for storing rules and I fully expect many rules to be added as needed.